### ISSUE TITLE
[Feature] Exposing a ChatController

### ISSUE THREAD

**Issue by DominikStarke** [flutter/ai#39]
I think it'd be great if the chat exposed a controller for more flexible handling of the following scenarios:

* Attaching files (For example ClipboardData) - right now this is possible with a custom messageSender, but the files aren't visible in the chat before submission
* Call sendMessage programmatically. For example there might be predefined prompts, which should be executed on click.
* Manage focus of the Chats Input. Especially on Desktop there are Scenarios, where manual focus management might be required (Also the current approach behaves kinda weird)

**Comment by csells** [comment]
> * Attaching files (For example ClipboardData) - right now this is possible with a custom messageSender, but the files aren't visible in the chat before submission

Can you give me an example of what you'd like to accomplish here?

> * Call sendMessage programmatically. For example there might be predefined prompts, which should be executed on click.

You can do that today with [the `suggestions` parameter](https://pub.dev/documentation/flutter_ai_toolkit/latest/flutter_ai_toolkit/LlmChatView/suggestions.html) of the `LlmChatView` constructor.

> * Manage focus of the Chats Input. Especially on Desktop there are Scenarios, where manual focus management might be required (Also the current approach behaves kinda weird)

Please log any bugs along with repro samples against any issues you find with the current approach to focus.

What kind of scenario would you like to handle with manual focus management?

**Comment by DominikStarke** [comment]
> Can you give me an example of what you'd like to accomplish here?  

Well I'm filling in files from the Clipboard: https://github.com/DominikStarke/ogre
Right now I'm tracking them outside the chat window.

> You can do that today with [the suggestions parameter](https://pub.dev/documentation/flutter_ai_toolkit/latest/flutter_ai_toolkit/LlmChatView/suggestions.html) of the LlmChatView constructor.  

That might work. What I want to achieve is if the user has something in his clipboard, he can invoke some "quick-actions", like writing a summary, or similar.

> Please log any bugs along with repro samples against any issues you find with the current approach to focus.  

I'll upload a video in a different issue. Soft-keyboards can't be hidden on mobile and on navigation once the keyboard loses focus and regains focus, just to be then dropped again, because the route is now out of focus. It makes for a very unpleasent route transition. That's on ios at least.

> What kind of scenario would you like to handle with manual focus management?  

When running on desktop my app has a simple hotkey, which brings it to front. When this happens I want to focus the input field (like in the video of the repo above). However right now it won't do that if the input isn't focused before the application is hidden.

**Comment by csells** [comment]
> > Can you give me an example of what you'd like to accomplish here?
> 
> Well I'm filling in files from the Clipboard: [DominikStarke/ogre](https://github.com/DominikStarke/ogre?rgh-link-date=2024-12-26T21%3A20%3A13Z) Right now I'm tracking them outside the chat window.

I'm sorry. I don't understand. Can you frame it in user actions, the results you'd like to provide to them and how you'd like the AIT to help with that?

> > What kind of scenario would you like to handle with manual focus management?
> 
> When running on desktop my app has a simple hotkey, which brings it to front. When this happens I want to focus the input field (like in the video of the repo above). However right now it won't do that if the input isn't focused before the application is hidden.

Great. If you could log a feature request for that functionality, that would be great.

**Comment by DominikStarke** [comment]
Let's say the user goes to a website, this page right here for example.
He sees there's already quite a lot of conversation going on and instead of reading everything he then:
* Selects all the text on the website
* Copies the selection to his clipboard
* Opens my app
* Now the app has the copied text as Context for the chat.

So far so good, that works already.
However I've had to create my own widget to visualize these kind of attachments.

Ideally I'd like them in the same area as the attachments opened with the file dialog (right over the chat input).


I hope this makes it more clear.

Here's a short Video:

https://github.com/user-attachments/assets/654b3c31-82c6-4ddb-8b25-fc52a152fc6e

**Comment by csells** [comment]
so, so your request to add support for the clipboard as an attachment, either manually or automatically? or is it for the app to be able to define custom attachment types?

**Comment by DominikStarke** [comment]
I'm lookiong for a way to add these attachments to the chat windows attached files, like normal attachments:
<img width="463" alt="Bildschirmfoto 2024-12-27 um 09 06 50" src="https://github.com/user-attachments/assets/bab34fc0-2418-432d-94d6-b2cef72c6597" />

I don't expect the package itself to handle clipboard data.

**Comment by csells** [comment]
have you tried doing this with a messageSender?

**Comment by DominikStarke** [comment]
Yes, I mentioned that earlier. Submission with messageSender works (like a charm) and the attachments are displayed in the chat history correctly.
My issue is rather... reducing visual clutter:  

<img width="471" alt="Bildschirmfoto 2024-12-28 um 09 18 44" src="https://github.com/user-attachments/assets/01eb6e5a-108f-4189-92f0-23a1cfb4dfe2" />

**Comment by csells** [comment]
understood. how should this work programatically?

**Comment by DominikStarke** [comment]
### Easy solution  
Probably an `attachments` parameter to `LlmChatView`?

```
final controller = LlmChatViewController();
LlmChatView(
  attachments: [...],
  provider: EchoProvider()
);
```

### Harder solution  

I wanted to avoid calling `setState` in an `LlmChatView` parent however, so that's where the "controller" idea comes from...

```
final controller = LlmChatViewController();
LlmChatView(
  controller: controller, // Optional
  provider: EchoProvider()
);
```

We could then have something like `controller.addAttachment(...)`, which would internally notify listening Widgets.
Same for requesting focus `controller.requestFocus()`.
I also think `controller.addSuggestions(...)`, `controller.setHistory(...)` would make sense, because otherwise its always necessary to call `setState` on the parent?

Or am I missing something?

**Comment by csells** [comment]
@DominikStarke can you provide a PR? I'm getting lost in the details here. Thanks.

**Comment by fanhaofei1234** [comment]
> [@DominikStarke](https://github.com/DominikStarke) can you provide a PR? I'm getting lost in the details here. Thanks.

I'd like to send messages without chat_input, but I can't right now

**Comment by csells** [comment]
yeah. I've run into that use case myself. It requires a controller or a state key, but is otherwise straightforward.

**Comment by csells** [comment]
@DominikStarke given either a controller or a state object, would being able to call `sendMessage(myPrompt, myAttachments)` serve your need? Or do you need to be able to set the prompt and the attachments as well?

**Comment by DominikStarke** [comment]
Hey @csells,
sorry for only responding now. I got frustrated with the openwebui api changing on every release and decided I need a break from my ai projects. I haven't been following this project very closely as a result.  

To answer your question:
sendMessage(myPrompt, myAttachments) might \*guess\* work. An application is then able to manage attachments outside / independent of chatView?

**Comment by csells** [comment]
that's my question. can you walk me through the UX you'd like to provide to your users? that will help me understand what's needed here.

**Comment by csells** [comment]
@DominikStarke feel free to re-open this if it comes up again for you.

**Comment by sunrimii** [comment]
@csells  It'd be nice to have a controller to programmatically set the input. For example, if a user selects some text and presses "Ask AI" from a context menu, it should automatically go to the chat screen and put the selected text into the chat input.

**Comment by csells** [comment]
I think that's an excellent feature request and I've put it on the list for v2. Thank yoU!
